Framework/Algorithms/src/FindPeaksConvolve.cpp

```diff
@@ -0,0 +1,376 @@
+    throw std::invalid_argument("Validation error: " + validation_res.first);
+  }
+  PARALLEL_FOR_IF(Kernel::threadSafe(*m_inputDataWS))
+  for (int i = 0; i < static_cast<int>(m_specCount); i++) {
+    performConvolution(i);
+  }
+  outputResults();
+}
+
+void FindPeaksConvolve::initiateValidators() {
+  m_validators.reserve(2); // Update depending upon number of validators.
+
+  auto mustBeNonNegative{std::make_shared<Kernel::BoundedValidator<int>>()};
+  mustBeNonNegative->setLower(0);
+  m_validators.emplace(std::make_pair("mustBeNonNegative", std::move(mustBeNonNegative)));
+
+  auto mustBeGreaterThanZero{std::make_shared<Kernel::BoundedValidator<double>>()};
+  mustBeGreaterThanZero->setLowerExclusive(0.0);
+  m_validators.emplace("mustBeGreaterThanZero", std::move(mustBeGreaterThanZero));
+
+  auto mustBeGreaterThanOne{std::make_shared<Kernel::BoundedValidator<int>>()};
+  mustBeGreaterThanOne->setLowerExclusive(1);
+  m_validators.emplace("mustBeGreaterThanOne", std::move(mustBeGreaterThanOne));
+}
+
+std::pair<std::string, int> FindPeaksConvolve::secondaryValidation() const {
+  std::string err_str{};
+  int err_code{0};
+
+  err_str += validatePeakExtentInput();
+
+  if (err_str != "") {
+    err_code = 1;
+  }
+  return std::make_pair(err_str, err_code);
+}
+
+const std::string FindPeaksConvolve::validatePeakExtentInput() const {
+  std::string err_str{};
+  const double peakExtent = getProperty("EstimatedPeakExtent");
+  const int peakExtentNBins = getProperty("EstimatedPeakExtentNBins");
+  if (peakExtent != EMPTY_DBL() && peakExtentNBins != EMPTY_INT()) {
+    err_str += "Peak Extent has been given in x units and in number of bins. Please specify one or the other. ";
+  }
+
+  return err_str;
+}
+
+void FindPeaksConvolve::storeClassProperties() {
+  m_inputDataWS = getProperty("InputWorkspace");
+  m_createIntermediateWorkspaces = getProperty("CreateIntermediateWorkspaces");
+  m_findHighestDatapointInPeak = getProperty("FindHighestDataPointInPeak");
+  m_iOverSigmaThreshold = getProperty("IOversigmaThreshold");
+  m_performBinaryClosing = getProperty("PerformBinaryClosing");
+
+  int wsIndex = getProperty("WorkspaceIndex");
+  if (wsIndex == EMPTY_INT()) {
+    m_specCount = static_cast<int>(m_inputDataWS->getNumberHistograms());
+    m_specNums.resize(m_specCount);
+    std::iota(std::begin(m_specNums), std::end(m_specNums), 0);
+  } else {
+    m_specCount = 1;
+    m_specNums.push_back(wsIndex);
+  }
+  m_peakResults.resize(m_specCount);
+}
+
+void FindPeaksConvolve::performConvolution(const size_t dataIndex) {
+  const auto specNum{m_specNums[dataIndex]};
+  const HistogramData::HistogramX *xData{&m_inputDataWS->x(specNum)};
+  const auto kernelBinCount{getKernelBinCount(xData)};
+  const Tensor1D kernel{createKernel(static_cast<int>(kernelBinCount))};
+
+  const auto binCount{m_inputDataWS->getNumberBins(specNum)};
+  const double paddingSize = (std::ceil(static_cast<double>(kernelBinCount) * 1.5) - 2) / 2;
```

Comment:
I think some comment required here about why we are padding, and why this size of padding is required